nixos/modules/tasks/network-interfaces-systemd.nix

```diff
@@ -53,20 +59,27 @@ in
     systemd.network =
       let
         domains = cfg.search ++ (optional (cfg.domain != null) cfg.domain);
-        genericNetwork = override:
-          let gateway = optional (cfg.defaultGateway != null) cfg.defaultGateway.address
+        defaultGateways = optional (cfg.defaultGateway != null) cfg.defaultGateway.address
             ++ optional (cfg.defaultGateway6 != null) cfg.defaultGateway6.address;
-          in {
+        genericNetwork = override:
+          {
             DHCP = override (dhcpStr cfg.useDHCP);
-          } // optionalAttrs (gateway != [ ]) {
-            gateway = override gateway;
+          } // optionalAttrs (defaultGateways != [ ]) {
+            gateway = override defaultGateways;
           } // optionalAttrs (domains != [ ]) {
             domains = override domains;
+          } // optionalAttrs (! cfg.enableIPv6) {
+            # if IPv6 is disabled, we shoud define additionnal attributes
+            # see https://github.com/coreos/bugs/issues/1419
+            networkConfig.LinkLocalAddressing = "no";
+            networkConfig.IPv6AcceptRA = "no";
           };
       in mkMerge [ {
         enable = true;
-        networks."99-main" = genericNetwork mkDefault;
       }
+      (mkIf (domains  != [ ] || defaultGateways != [ ] || cfg.useDHCP)  {
```

Comment:
Does that mean we won't create a default network (read: we won't do ANY configuration) without DHCP, search domain or an Gateway?

Out of my hea: If I use SLAAC with IPv6 I MUST set `useDHCP` or else I do not have a `.network` file (and thus no networking)?

Reply:
The problem I had, is that this `99-main` network file was messing my setup (again, without IPv6). For example, my tunnel interfaces (from openvpn) where rendered unusable because networkd was messing with them. Also, I had the `configuring` problems as well.

To cover the problem of ipv6, maybe we add `cfg.enableIPv6` to the conditionnal ?

Reply:
We are gaining a lot of complexity in the networking configuration. If you think that would solve the issue go for it.

I would like to see tests for these kind of things so we can verify that it is working as intended vs. having humans parse nix expressions :-)

Reply:
I also don't like this implementation very much.

We can use the following to prevent networkd from messing with all the interfaces :
```
[Link]
Unmanaged=yes
```

What do you think about the following solution: we add a global parameter `networking.acceptRA`, if `networking.useDHCP=false` and `networking.enableIPv6=false` and `networking.acceptRA=false` we add the unmanaged flag in the "99-main" unit.

I think this is simpler, and enables use of domains and defaultGateway without messing with interfaces (e.g. tunnels) when it's not needed. I will test this if you think this is better.